### PR TITLE
Extensible org.zaproxy.zap.model.Tech

### PR DESCRIPTION
**Pull request by t4skforce**
As discussed today on group discussion, the PR for adding handling of dynamically changed Tech.

* Added support for dynamic techlist in UI
* adapted tests

Example use in Javascript plugin
`
var Tech = Java.type('org.zaproxy.zap.model.Tech');
Tech.add(new Tech(Tech.SCM, 'Murcurial'))
Tech.remove(Tech.get('Tomcat'))
`

TODO: Unittests for new functions

### CODE REVIEW

File: zap/src/main/java/org/zaproxy/zap/model/Tech.java [hunk 1]
```diff
@@ -232,7 +232,7 @@ public int compareTo(Tech o) {
      */
     public static void add(Tech tech) {
         Assert.isNonEmpty(tech);
-        if (tech.getParent() != null) {
+        if (tech.getParent() == null) {
```
Comment (kingthorin) [inline zap/src/main/java/org/zaproxy/zap/model/Tech.java:235]: Its going to add null techs ?

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

To address the build failure the code needs formatting via the spotlessApply grade task.

File: zap/src/main/java/org/zaproxy/zap/model/Tech.java (hunk 1 above)
Comment (t4skforce) [reply zap/src/main/java/org/zaproxy/zap/model/Tech.java:235]: No it isn't that one is a check if we have a top-level/root tech element (no parent = new root)

Comment (t4skforce) [reply zap/src/main/java/org/zaproxy/zap/model/Tech.java]: I fixed the null check, and Tech with parent null, means top level tech like Tech.Db.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/AbstractPlugin.java [hunk 2]
```diff
@@ -296,7 +296,7 @@ protected void sendAndReceive(
 
         if (this.getDelayInMs() > 0) {
             try {
-                Thread.sleep(this.getDelayInMs());
```
Comment (psiinon) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/AbstractPlugin.java:299]: Changing things like this is fine but would be much better in another 'clean up' PR - it detracts from the important changes.

File: zap/src/main/java/org/parosproxy/paros/network/HttpRequestHeader.java [hunk 3]
```diff
@@ -56,6 +56,8 @@
 // ZAP: 2019/06/01 Normalise line endings.
 // ZAP: 2019/06/05 Normalise format/style.
 // ZAP: 2019/12/09 Address deprecation of getHeaders(String) Vector method.
+// ZAP: 2020/11/10 Add convenience method isCss(), refactor isImage() to use new private method
+// isSpecificType(Pattern).
```
Comment (psiinon) [inline zap/src/main/java/org/parosproxy/paros/network/HttpRequestHeader.java:60]: Looks like you've pulled in other changes :(

**Review by psiinon — CHANGES_REQUESTED** [review]
Looks like you've made a load of 'cleanliness' changes and pulled other PRs as well. This makes this PR much harder to review. Please just include the functional changes required.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/AbstractPlugin.java (hunk 2 above)
Comment (t4skforce) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/AbstractPlugin.java:299]: true, I took the boy-scout approach on the files I changed a lot

File: zap/src/main/java/org/parosproxy/paros/network/HttpRequestHeader.java (hunk 3 above)
Comment (t4skforce) [reply zap/src/main/java/org/parosproxy/paros/network/HttpRequestHeader.java:60]: :( this sign-off thing is driving me nuts. I think I have to reopen the PR from a new clean bunch

**Pull request closed without merging**